taichi/runtime/llvm/runtime.cpp

```diff
@@ -926,9 +926,67 @@ void clear_list(LLVMRuntime *runtime, StructMeta *parent, StructMeta *child) {
  * The element list of a SNode, maintains pointers to its instances, and
  * instances' parents' coordinates
  */
-void element_listgen(LLVMRuntime *runtime,
-                     StructMeta *parent,
-                     StructMeta *child) {
+
+// For the root node there is only one container,
+// therefore we use a special kernel for more parallelism.
+void element_listgen_root(LLVMRuntime *runtime,
+                          StructMeta *parent,
+                          StructMeta *child) {
+  // If there's just one element in the parent list, we need to use the blocks
+  // (instead of threads) to split the parent container
+  auto parent_list = runtime->element_lists[parent->snode_id];
+  auto child_list = runtime->element_lists[child->snode_id];
+  // Cache the func pointers here for better compiler optimization
+  auto parent_refine_coordinates = parent->refine_coordinates;
+  auto parent_is_active = parent->is_active;
+  auto parent_lookup_element = parent->lookup_element;
+  auto child_get_num_elements = child->get_num_elements;
+  auto child_from_parent_element = child->from_parent_element;
+#if ARCH_cuda
+  // All blocks share the only root container, which has only one child
+  // container.
+  int j_start = block_idx();
+  int j_step = grid_dim();
+  // Each thread processes a subset of the child container for more parallelism.
+  int c_start = thread_idx();
+  int c_step = block_dim();
+#else
+  int j_start = 0;
+  int j_step = 1;
+  int c_start = 0;
+  int c_step = 1;
+#endif
+  auto element = parent_list->get<Element>(0);
+
+  int j_lower = element.loop_bounds[0] + j_start;
+  int j_higher = element.loop_bounds[1];
+  for (int j = j_lower; j < j_higher; j += j_step) {
```

Comment:
Do we actually need this loop, given that `j_higher` is always 1 (IIUC..)?

Reply:
Thanks for pointing this out - I added some comments there
```
  // Note that the root node has only one container, and the `element`
  // representing that single container has only one 'child':
  // element.loop_bounds[0] = 0 and element.loop_bounds[1] = 1
  // Therefore, compared with element_listgen_nonroot,
  // we need neither `i` to loop over the `elements`, nor `j` to
  // loop over the children.
```